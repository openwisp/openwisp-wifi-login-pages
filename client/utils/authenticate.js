import {sessionStorage} from "./storage";

const authenticate = (cookies, orgSlug) => {
  const token = cookies.get(`${orgSlug}_auth_token`);
  const sessionKey = sessionStorage.getItem(`${orgSlug}_auth_token`);
  if (sessionKey) {
    cookies.remove(`${orgSlug}_auth_token`, {path: "/"});
    cookies.remove(`${orgSlug}_username`, {path: "/"});
    return true;
  }
  if (token) return true;
  return false;
};
export default authenticate;
