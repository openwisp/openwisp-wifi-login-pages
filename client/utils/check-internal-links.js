const isInternalLink = (link) => {
  if (link.length && link[0] === "/") return true;
  return false;
};
export default isInternalLink;
