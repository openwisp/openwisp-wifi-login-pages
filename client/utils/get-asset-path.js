const getAssetPath = (slug, relativePath) => `/assets/${slug}/${relativePath}`;
export default getAssetPath;
