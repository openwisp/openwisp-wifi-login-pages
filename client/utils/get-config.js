const loadConfig = (slug) => import(`../configs/${slug}.json`);
const getConfig = async (slug) => {
  let config;
  try {
    config = await loadConfig(slug);
  } catch (err) {
    console.error(`Configuration for organization ${slug} does not exists.`);
    config = undefined;
  }
  return config;
};

export default getConfig;
