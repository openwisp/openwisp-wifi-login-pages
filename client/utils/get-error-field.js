import React from "react";

const getErrorField = (errors, key, nonField = false, className = "") =>
  errors[key] ? (
    <div className={!nonField ? `error ${className}` : "error non-field"}>
      <span className="icon">!</span>
      <span className="text">{errors[key]}</span>
    </div>
  ) : null;

export default getErrorField;
