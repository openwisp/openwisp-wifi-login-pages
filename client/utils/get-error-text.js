const getErrorText = (error, fallbackVal = "") => {
  const {data} = error.response;
  // eslint-disable-next-line no-nested-ternary
  return data.detail
    ? data.detail
    : data.non_field_errors
    ? data.non_field_errors[0]
    : fallbackVal;
};
export default getErrorText;
