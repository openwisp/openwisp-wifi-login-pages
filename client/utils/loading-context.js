import React from "react";

export const loadingContextValue = {setLoading: () => {}, getLoading: () => {}};
const LoadingContext = React.createContext(loadingContextValue);

export default LoadingContext;
