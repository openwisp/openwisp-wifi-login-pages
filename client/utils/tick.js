const tick = function () {
  return new Promise((resolve) => {
    process.nextTick(resolve);
  });
};

export default tick;
