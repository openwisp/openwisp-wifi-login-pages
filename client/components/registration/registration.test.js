/* eslint-disable prefer-promise-reject-errors */
/* eslint-disable camelcase */
import axios from "axios";
import {shallow} from "enzyme";
import React from "react";
import {toast} from "react-toastify";
import PropTypes from "prop-types";
import {Route} from "react-router-dom";
import PhoneInput from "react-phone-input-2";
import {t} from "ttag";
import Modal from "../modal";
import {loadingContextValue} from "../../utils/loading-context";
import tick from "../../utils/tick";

import getConfig from "../../utils/get-config";
import loadTranslation from "../../utils/load-translation";
import Registration from "./registration";
import submitOnEnter from "../../utils/submit-on-enter";
import PasswordToggleIcon from "../../utils/password-toggle";
import mountComponent from "./test-utils";
import InfoModal from "../../utils/modal";
import history from "../../utils/history";

jest.mock("../../utils/get-config");
jest.mock("../../utils/load-translation");
jest.mock("../../utils/submit-on-enter");
jest.mock("../../utils/history");
jest.mock("axios");

const createTestProps = (props, configName = "default") => {
  const config = getConfig(configName);
  return {
    language: "en",
    orgSlug: configName,
    orgName: config.name,
    settings: config.settings,
    registration: config.components.registration_form,
    privacyPolicy: config.privacy_policy,
    termsAndConditions: config.terms_and_conditions,
    authenticate: jest.fn(),
    setTitle: jest.fn(),
    setUserData: jest.fn(),
    loading: false,
    match: {
      path: "default/registration",
    },
    ...props,
  };
};

const responseData = {
  key: "8a2b2b2dd963de23c17db30a227505f879866630",
  radius_user_token: "Lbdh3GKD7hvXUS5NUu5yoE4x5fCPPqlsXo7Ug8ld",
};

describe("<Registration /> rendering with placeholder translation tags", () => {
  const props = createTestProps();
  const wrapper = shallow(<Registration {...props} />, {
    context: loadingContextValue,
  });
  it("should render translation placeholder correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("<Registration /> rendering", () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    loadTranslation("en", "default");
  });
  it("should render correctly", () => {
    props = createTestProps();
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
    });
    expect(wrapper).toMatchSnapshot();
  });
});

describe("<Registration /> interactions", () => {
  let props;
  let wrapper;
  let originalError;
  let lastConsoleOutuput;

  beforeEach(() => {
    originalError = console.error;
    lastConsoleOutuput = null;
    console.error = (data) => {
      lastConsoleOutuput = data;
    };
    props = createTestProps();
    Registration.contextTypes = {
      setLoading: PropTypes.func,
      getLoading: PropTypes.func,
    };
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
    });
  });
  afterEach(() => {
    console.error = originalError;
  });
  it("should change state values when handleChange function is invoked", () => {
    wrapper
      .find(".row.email input")
      .simulate("change", {target: {value: "test email", name: "email"}});
    expect(wrapper.state("email")).toEqual("test email");
    wrapper
      .find(".row.password input")
      .simulate("change", {target: {value: "testpassword", name: "password1"}});
    expect(wrapper.state("password1")).toEqual("testpassword");
    wrapper
      .find(".row.password-confirm input")
      .simulate("change", {target: {value: "testpassword", name: "password2"}});
    expect(wrapper.state("password2")).toEqual("testpassword");
  });

  it("should execute handleSubmit correctly when form is submitted", () => {
    axios
      .mockImplementationOnce(() =>
        Promise.reject({
          response: {
            data: {
              email: "email error",
              detail: "nonField error",
              password1: "password1 error",
            },
          },
        }),
      )
      .mockImplementationOnce(() =>
        Promise.reject({
          status: 500,
          statusText: "Internal server error",
          response: {
            data: {
              detail: "Internal server error",
            },
          },
        }),
      )
      .mockImplementationOnce(() =>
        Promise.reject({
          status: 504,
          statusText: "Gateway Timeout",
          response: {
            data: {},
          },
        }),
      )
      .mockImplementationOnce(() =>
        Promise.resolve({
          status: 201,
          statusText: "CREATED",
          data: responseData,
        }),
      )
      .mockImplementationOnce(() =>
        Promise.reject({
          status: 400,
          statusText: "Bad Request",
          response: {
            data: {
              billing_info: {
                billingError: "registration error",
              },
            },
          },
        }),
      );
    wrapper.setState({
      password1: "wrong password",
      password2: "wrong password1",
    });
    const event = {preventDefault: () => {}};
    const spyToast = jest.spyOn(toast, "error");
    wrapper.instance().handleSubmit(event);
    expect(
      wrapper.update().find(".row.password-confirm div.error"),
    ).toHaveLength(1);
    expect(
      wrapper.update().find(".row.password-confirm input.error"),
    ).toHaveLength(1);
    wrapper.setState({
      password1: "password",
      password2: "password",
    });
    wrapper.setProps({
      registration: {
        ...props.registration,
        input_fields: {
          ...props.registration.input_fields,
        },
      },
    });
    return wrapper
      .instance()
      .handleSubmit(event)
      .then(() => {
        expect(wrapper.instance().state.errors).toEqual({
          birth_date: "",
          city: "",
          country: "",
          email: "email error",
          first_name: "",
          last_name: "",
          location: "",
          password1: "password1 error",
          password2: "",
          street: "",
          tax_number: "",
          username: "",
          zipcode: "",
        });
        expect(wrapper.find("div.error")).toHaveLength(2);
        expect(wrapper.instance().props.authenticate.mock.calls.length).toBe(0);
        expect(lastConsoleOutuput).not.toBe(null);
        expect(spyToast.mock.calls.length).toBe(1);
        lastConsoleOutuput = null;
      })
      .then(() =>
        wrapper
          .instance()
          .handleSubmit(event)
          .then(() => {
            expect(
              wrapper.instance().props.authenticate.mock.calls.length,
            ).toBe(0);
            expect(lastConsoleOutuput).not.toBe(null);
            expect(spyToast.mock.calls.length).toBe(2);
            lastConsoleOutuput = null;
          }),
      )
      .then(() =>
        wrapper
          .instance()
          .handleSubmit(event)
          .then(() => {
            expect(
              wrapper.instance().props.authenticate.mock.calls.length,
            ).toBe(0);
            expect(lastConsoleOutuput).not.toBe(null);
            expect(spyToast.mock.calls.length).toBe(3);
            lastConsoleOutuput = null;
          }),
      )
      .then(() =>
        wrapper
          .instance()
          .handleSubmit(event)
          .then(() => {
            expect(wrapper.instance().state.errors).toEqual({});
            expect(wrapper.instance().state.success).toEqual(true);
            expect(wrapper.find(".success")).toHaveLength(1);
            expect(
              wrapper.instance().props.authenticate.mock.calls.length,
            ).toBe(1);
            expect(lastConsoleOutuput).toBe(null);
            expect(spyToast.mock.calls.length).toBe(3);
            lastConsoleOutuput = null;
          }),
      )
      .then(() =>
        wrapper
          .instance()
          .handleSubmit(event)
          .then(() => {
            expect(
              wrapper.instance().props.authenticate.mock.calls.length,
            ).toBe(1);
            expect(lastConsoleOutuput).not.toBe(null);
            expect(spyToast.mock.calls.length).toBe(4);
            lastConsoleOutuput = null;
          }),
      );
  });
  it("test optional fields disabled", async () => {
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    expect(wrapper.find(".first_name").length).toEqual(0);
    expect(wrapper.find(".last_name").length).toEqual(0);
    expect(wrapper.find(".birth_date").length).toEqual(0);
    expect(wrapper.find(".location").length).toEqual(0);
  });
  it("test optional fields allowed", async () => {
    props.registration.input_fields.first_name.setting = "allowed";
    props.registration.input_fields.location.setting = "allowed";
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    expect(wrapper.find("[htmlFor='first_name']").text()).toEqual(
      "First name (optional)",
    );
    expect(wrapper.find("[htmlFor='location']").text()).toEqual(
      "Location (optional)",
    );
    expect(wrapper.find(".last_name").length).toEqual(0);
    expect(wrapper.find(".birth_date").length).toEqual(0);
  });
  it("test optional fields mandatory", async () => {
    props.registration.input_fields.birth_date.setting = "mandatory";
    props.registration.input_fields.first_name.setting = "mandatory";
    props.registration.input_fields.last_name.setting = "allowed";
    props.registration.input_fields.location.setting = "allowed";
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    expect(wrapper.find("[htmlFor='first_name']").text()).toEqual("First name");
    expect(wrapper.find("[htmlFor='birth_date']").text()).toEqual("Birth date");
    expect(wrapper.find("[htmlFor='last_name']").text()).toEqual(
      "Last name (optional)",
    );
    expect(wrapper.find("[htmlFor='location']").text()).toEqual(
      "Location (optional)",
    );
  });
  it("should execute authenticate in mobile phone verification flow", async () => {
    axios.mockImplementationOnce(() =>
      Promise.resolve({
        status: 201,
        statusText: "CREATED",
        data: responseData,
      }),
    );
    props.settings = {mobile_phone_verification: true};
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    const event = {preventDefault: () => {}};
    const errorSpyToast = jest.spyOn(toast, "error");
    wrapper.setState({
      password1: "password",
      password2: "password",
    });
    wrapper.instance().handleSubmit(event);
    await tick();
    expect(wrapper.instance().state.errors).toEqual({});
    expect(wrapper.instance().state.success).toEqual(true);
    expect(wrapper.find(".success")).toHaveLength(1);
    expect(wrapper.instance().props.authenticate.mock.calls.length).toBe(1);
    expect(errorSpyToast.mock.calls.length).toBe(4);
    const setUserDataMock = wrapper.instance().props.setUserData.mock;
    expect(setUserDataMock.calls.length).toBe(1);
    expect(setUserDataMock.calls.pop()).toEqual([
      {is_verified: false, auth_token: responseData.key},
    ]);
  });
  it("should toggle modal", async () => {
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    expect(wrapper.instance().state.modalActive).toEqual(false);
    wrapper.instance().toggleModal();
    expect(wrapper.instance().state.modalActive).toEqual(true);
  });
  it("should show modal if user is already registered with other organizations", async () => {
    const data = {
      detail: "user already registered",
      organizations: [{name: "default", slug: "default"}],
    };
    axios.mockImplementationOnce(() =>
      Promise.reject({
        response: {
          status: 409,
          statusText: "CONFLICT",
          data,
        },
      }),
    );
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    jest.spyOn(wrapper.instance(), "toggleModal");
    jest.spyOn(wrapper.instance(), "handleResponse");
    wrapper.instance().setState({
      password1: "testpassword",
      password2: "testpassword",
      email: "tester@openwisp.org",
    });
    const event = {preventDefault: () => {}};
    expect(wrapper.instance().state.errors).toEqual({});
    wrapper.instance().handleSubmit(event);
    await tick();
    expect(wrapper.instance().state.errors).toEqual(data);
    expect(wrapper.instance().toggleModal).toHaveBeenCalled();
    const modalWrapper = wrapper.find(InfoModal).shallow();
    expect(modalWrapper).toMatchSnapshot();
    modalWrapper
      .find(".modal-buttons button:first-child")
      .simulate("click", {});
    expect(wrapper.instance().handleResponse).toHaveBeenCalledWith(true);
    modalWrapper.find(".modal-buttons button:last-child").simulate("click", {});
    expect(wrapper.instance().handleResponse).toHaveBeenCalledWith(false);
  });
  it("should execute handleResponse correctly", async () => {
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    jest.spyOn(wrapper.instance(), "toggleModal");
    const spyToast = jest.spyOn(toast, "info");
    wrapper.instance().handleResponse(true);
    expect(spyToast).toHaveBeenCalledWith(t`PLEASE_LOGIN`);
    expect(history.push).toHaveBeenCalledWith("/default/login");
    wrapper.instance().handleResponse(false);
    expect(wrapper.instance().toggleModal).toHaveBeenCalled();
  });
  it("should show modal if user is registered but not associated with any org", async () => {
    const data = {
      detail: "user already registered",
      organizations: [],
    };
    axios.mockImplementationOnce(() =>
      Promise.reject({
        response: {
          status: 409,
          statusText: "CONFLICT",
          data,
        },
      }),
    );
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    const event = {preventDefault: () => {}};
    wrapper.instance().handleSubmit(event);
    await tick();
    expect(wrapper.instance().state.errors).toEqual(data);
    const modalWrapper = wrapper.find(InfoModal).shallow();
    expect(modalWrapper.contains(<p>{t`NO_ORGS`}</p>)).toBe(true);
    expect(modalWrapper).toMatchSnapshot();
  });
});

describe("Registration and Mobile Phone Verification interactions", () => {
  let props;
  let wrapper;
  const event = {preventDefault: jest.fn()};

  beforeEach(() => {
    props = createTestProps({}, "test-org-2");
    props.configuration = getConfig("test-org-2");
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it("should show phone number field", async () => {
    wrapper = await mountComponent(props);
    expect(wrapper.find("input[name='phone_number']").length).toBe(1);
  });

  it("should render PhoneInput lazily and handlers should work correctly", async () => {
    wrapper = shallow(<Registration {...props} />);
    const spyFn = jest.fn();
    wrapper.instance().handleChange = spyFn;
    const component = wrapper.find("Suspense");
    expect(component).toMatchSnapshot();
    expect(component.find("lazy").length).toBe(1);
    const prop = component.find("lazy").props();
    expect(prop).toEqual({
      country: "it",
      enableSearch: false,
      excludeCountries: [],
      inputProps: {
        autoComplete: "tel",
        className: "form-control input ",
        id: "phone-number",
        name: "phone_number",
        required: true,
      },
      name: "phone_number",
      onChange: expect.any(Function),
      onKeyDown: expect.any(Function),
      onlyCountries: [],
      placeholder: "enter mobile phone number",
      preferredCountries: [],
      value: "",
    });
    prop.onChange("+911234567890");
    expect(spyFn).toHaveBeenCalledWith({
      target: {
        name: "phone_number",
        value: "++911234567890",
      },
    });
    component.find("lazy").props().onKeyDown({});
    expect(submitOnEnter.mock.calls.length).toEqual(1);
    expect(submitOnEnter.mock.calls.pop()).toEqual([
      {},
      expect.any(Object),
      "registration-form",
    ]);
  });

  it("should process successfully", async () => {
    wrapper = await mountComponent(props);
    expect(wrapper.find("input[name='phone_number']").length).toBe(1);
    expect(wrapper.find("form")).toHaveLength(1);
    const component = wrapper.find(Registration).instance();
    const handleSubmit = jest.spyOn(component, "handleSubmit");

    axios.mockImplementationOnce(() =>
      Promise.resolve({
        status: 201,
        statusText: "CREATED",
        data: responseData,
      }),
    );

    wrapper.find("input[name='phone_number']").simulate("change", {
      target: {value: "+393660011333", name: "phone_number"},
    });
    wrapper.find("input[name='email']").simulate("change", {
      target: {value: "tester@openwisp.io", name: "email"},
    });
    wrapper
      .find("input[name='password1']")
      .simulate("change", {target: {value: "tester123", name: "password1"}});
    wrapper
      .find("input[name='password2']")
      .simulate("change", {target: {value: "tester123", name: "password2"}});
    wrapper.find("form").simulate("submit", event);
    await tick();
    const registration = wrapper.find(Registration).instance();
    expect(registration.state.errors).toEqual({});
    expect(handleSubmit).toHaveBeenCalled();
    expect(event.preventDefault).toHaveBeenCalled();
    const setUserDataMock = registration.props.setUserData.mock;
    expect(setUserDataMock.calls.length).toBe(1);
    expect(setUserDataMock.calls.pop()).toEqual([
      {is_verified: false, auth_token: responseData.key},
    ]);
  });
  it("should load fallback before PhoneInput and handlers should work correctly", async () => {
    wrapper = shallow(<Registration {...props} />);
    const handleChange = jest.spyOn(wrapper.instance(), "handleChange");
    const component = wrapper.find("Suspense");
    const {fallback} = component.props();
    expect(fallback.type).toEqual("input");
    expect(fallback.props).toEqual({
      name: "phone_number",
      className: "input",
      value: "",
      onChange: expect.any(Function),
      onKeyDown: expect.any(Function),
      placeholder: "enter mobile phone number",
      type: "tel",
    });
    fallback.props.onChange("+911234567890");
    expect(handleChange).toHaveBeenCalledWith({
      target: {
        name: "phone_number",
        value: "++911234567890",
      },
    });
    fallback.props.onKeyDown({});
    expect(submitOnEnter.mock.calls.length).toEqual(1);
    expect(submitOnEnter.mock.calls.pop()).toEqual([
      {},
      expect.any(Object),
      "registration-form",
    ]);
  });
  it("should render modal", () => {
    props = createTestProps();
    wrapper = shallow(<Registration {...props} />);
    let pathMap = {};
    pathMap = wrapper.find(Route).reduce((mapRoute, route) => {
      const map = mapRoute;
      const routeProps = route.props();
      map[routeProps.path] = routeProps.render;
      return map;
    }, {});
    expect(pathMap["default/registration/:name"]).toEqual(expect.any(Function));
    const render = pathMap["default/registration/:name"];
    const Comp = React.createElement(Modal).type;
    expect(JSON.stringify(render({}))).toStrictEqual(JSON.stringify(<Comp />));
  });
  it("should send post data with optional fields", async () => {
    axios.mockImplementationOnce(() =>
      Promise.resolve({
        status: 201,
        statusText: "CREATED",
        data: responseData,
      }),
    );
    props = createTestProps();
    Registration.contextTypes = {
      setLoading: PropTypes.func,
      getLoading: PropTypes.func,
    };
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    wrapper.instance().setState({first_name: "OpenWISP"});
    wrapper.instance().handleSubmit(event);
    expect(axios).toHaveBeenCalledWith({
      data: {
        birth_date: "",
        email: "",
        first_name: "OpenWISP",
        last_name: "",
        location: "",
        password1: "",
        password2: "",
        username: "",
      },
      headers: {
        "content-type": "application/json",
        "accept-language": expect.any(String),
      },
      method: "post",
      url: "/api/v1/default/account/",
    });
  });
  it("should toggle password icon for both password fields in PasswordToggleIcon", async () => {
    wrapper = shallow(<Registration {...props} />, {
      context: loadingContextValue,
      disableLifecycleMethods: true,
    });
    const nodes = wrapper.find(PasswordToggleIcon);
    expect(nodes.length).toEqual(2);
    expect(nodes.at(0).props()).toEqual({
      hidePassword: true,
      inputRef: {current: null},
      isVisible: false,
      parentClassName: "",
      secondInputRef: {current: null},
      toggler: expect.any(Function),
    });
    expect(wrapper.instance().state.hidePassword).toEqual(true);
    nodes.at(0).props().toggler();
    expect(wrapper.instance().state.hidePassword).toEqual(false);
    expect(nodes.at(1).props()).toEqual({
      hidePassword: true,
      inputRef: {current: null},
      isVisible: false,
      parentClassName: "",
      secondInputRef: {current: null},
      toggler: expect.any(Function),
    });
    nodes.at(1).props().toggler();
    expect(wrapper.instance().state.hidePassword).toEqual(false);
  });
});

describe("Registration without identity verification (Email registration)", () => {
  let props;
  let wrapper;
  const event = {preventDefault: jest.fn()};

  beforeEach(() => {
    props = createTestProps({}, "test-org-2");
    props.configuration = getConfig("test-org-2");
    props.configuration.settings.mobile_phone_verification = false;
    props.configuration.settings.subscriptions = false;
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it("should not show phone number field", async () => {
    wrapper = await mountComponent(props);
    expect(wrapper.exists(PhoneInput)).toBe(false);
    expect(wrapper.find("form")).toHaveLength(1);
    expect(wrapper.find("input[name='phone_number']").length).toBe(0);
  });

  it("should process successfully", async () => {
    wrapper = await mountComponent(props);
    expect(wrapper.find("input[name='phone_number']").length).toBe(0);
    expect(wrapper.find("form")).toHaveLength(1);
    expect(wrapper.find("input[name='phone_number']").length).toBe(0);
    const component = wrapper.find(Registration).instance();
    const handleChange = jest.spyOn(component, "handleChange");
    const handleSubmit = jest.spyOn(component, "handleSubmit");

    axios.mockImplementationOnce(() =>
      Promise.resolve({
        status: 201,
        statusText: "CREATED",
        data: responseData,
      }),
    );

    wrapper.find("input[name='email']").simulate("change", {
      target: {value: "tester@openwisp.io", name: "email"},
    });
    wrapper
      .find("input[name='password1']")
      .simulate("change", {target: {value: "tester123", name: "password1"}});
    wrapper
      .find("input[name='password2']")
      .simulate("change", {target: {value: "tester123", name: "password2"}});
    wrapper.find("form").simulate("submit", event);
    await tick();
    expect(wrapper.find(Registration).instance().state.errors).toEqual({});
    expect(handleChange.mock.calls.length).toEqual(2);
    expect(handleSubmit).toHaveBeenCalled();
    expect(event.preventDefault).toHaveBeenCalled();
  });
  it("should set title", async () => {
    wrapper = await mountComponent(props);
    expect(wrapper.find("form")).toHaveLength(1);
    const component = wrapper.find(Registration);
    const setTitleMock = component.props().setTitle.mock;
    expect(setTitleMock.calls.pop()).toEqual(["Sign up", props.orgName]);
  });
  it("should set country when selectedCountry is executed", async () => {
    wrapper = await mountComponent(props);
    expect(wrapper.find("form")).toHaveLength(1);
    const component = wrapper.find(Registration);
    const data = {
      value: "India",
    };
    component.instance().selectedCountry(data);
    expect(component.instance().state.countrySelected).toEqual(data);
    expect(component.instance().state.country).toEqual(data.value);
  });
});
