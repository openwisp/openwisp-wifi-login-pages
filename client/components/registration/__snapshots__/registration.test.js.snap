// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> interactions should show modal if user is already registered with other organizations 1`] = `
<div
  className="modal is-visible"
>
  <div
    className="modal-container bg"
  >
    <button
      className="modal-close-btn"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      type="button"
    >
      ✖
    </button>
    <div>
      <div
        className="message"
      >
        <p>
          A user with credentials like the one being entered already exists in other organizations:
        </p>
        <ul>
          <li
            className="org-list"
            key="default"
          >
            default
          </li>
        </ul>
        <p>
          Do you own the account mentioned above and want to sign in with this organization?
        </p>
      </div>
    </div>
    <p
      className="modal-buttons"
    >
      <button
        className="button partial"
        onClick={[Function]}
        type="button"
      >
        Yes
      </button>
      <button
        className="button partial"
        onClick={[Function]}
        type="button"
      >
        No
      </button>
    </p>
  </div>
</div>
`;

exports[`<Registration /> interactions should show modal if user is registered but not associated with any org 1`] = `
<div
  className="modal is-visible"
>
  <div
    className="modal-container bg"
  >
    <button
      className="modal-close-btn"
      onClick={[Function]}
      type="button"
    >
      ✖
    </button>
    <div>
      <div
        className="message"
      >
        <p>
          A user with these credentials already exists.
        </p>
        <ul />
        <p>
          Do you own the account mentioned above and want to sign in with this organization?
        </p>
      </div>
    </div>
    <p
      className="modal-buttons"
    >
      <button
        className="button partial"
        onClick={[Function]}
        type="button"
      >
        Yes
      </button>
      <button
        className="button partial"
        onClick={[Function]}
        type="button"
      >
        No
      </button>
    </p>
  </div>
</div>
`;

exports[`<Registration /> rendering should render correctly 1`] = `
<Fragment>
  <div
    className="container content"
    id="registration"
  >
    <div
      className="inner"
    >
      <form
        className="main-column "
        id="registration-form"
        onSubmit={[Function]}
      >
        <div
          className="inner"
        >
          <div
            className="fieldset"
          >
            <div
              className="row email"
            >
              <label
                htmlFor="email"
              >
                Email
              </label>
              <input
                autoComplete="email"
                className="input "
                id="email"
                name="email"
                onChange={[Function]}
                pattern=".+@.+\\\\..+"
                placeholder="email address"
                required={true}
                title="enter valid email address"
                type="email"
                value=""
              />
            </div>
            <div
              className="row password"
            >
              <label
                htmlFor="password"
              >
                Password
              </label>
              <input
                autoComplete="new-password"
                className="input "
                id="password"
                name="password1"
                onChange={[Function]}
                pattern=".{6,}"
                placeholder="password"
                required={true}
                title="password must be a minimum of 6 characters"
                type="password"
                value=""
              />
              <PasswordToggleIcon
                hidePassword={true}
                inputRef={
                  Object {
                    "current": null,
                  }
                }
                isVisible={false}
                parentClassName=""
                secondInputRef={
                  Object {
                    "current": null,
                  }
                }
                toggler={[Function]}
              />
            </div>
            <div
              className="row password-confirm"
            >
              <label
                htmlFor="password-confirm"
              >
                Confirm password
              </label>
              <input
                autoComplete="new-password"
                className="input "
                id="password-confirm"
                name="password2"
                onChange={[Function]}
                pattern=".{6,}"
                placeholder="confirm password"
                required={true}
                title="password must be a minimum of 6 characters"
                type="password"
                value=""
              />
              <PasswordToggleIcon
                hidePassword={true}
                inputRef={
                  Object {
                    "current": null,
                  }
                }
                isVisible={false}
                parentClassName=""
                secondInputRef={
                  Object {
                    "current": null,
                  }
                }
                toggler={[Function]}
              />
            </div>
          </div>
          <div
            className="row add-info"
          >
            By signing up, you accept the 
            <Link
              className="link additional registration"
              key="terms-and-conditions"
              to="/default/registration/terms-and-conditions"
            >
              terms and conditions
            </Link>
             and the 
            <Link
              className="link additional registration"
              key="privacy-policy"
              to="/default/registration/privacy-policy"
            >
              privacy policy
            </Link>
             of this WiFi service.
          </div>
          <div
            className="row register"
          >
            <input
              className="button full"
              type="submit"
              value="Sign up"
            />
          </div>
          <div
            className="row links"
          >
            <p>
              <Link
                className="link"
                to="/default/login"
              >
                Already have an account? Log in
              </Link>
            </p>
          </div>
        </div>
      </form>
      <Connect(Contact) />
    </div>
  </div>
  <Route
    path="default/registration/:name"
    render={[Function]}
  />
</Fragment>
`;

exports[`<Registration /> rendering with placeholder translation tags should render translation placeholder correctly 1`] = `
<Fragment>
  <div
    className="container content"
    id="registration"
  >
    <div
      className="inner"
    >
      <form
        className="main-column "
        id="registration-form"
        onSubmit={[Function]}
      >
        <div
          className="inner"
        >
          <div
            className="fieldset"
          >
            <div
              className="row email"
            >
              <label
                htmlFor="email"
              >
                EMAIL
              </label>
              <input
                autoComplete="email"
                className="input "
                id="email"
                name="email"
                onChange={[Function]}
                pattern=".+@.+\\\\..+"
                placeholder="EMAIL_PHOLD"
                required={true}
                title="EMAIL_PTRN_DESC"
                type="email"
                value=""
              />
            </div>
            <div
              className="row password"
            >
              <label
                htmlFor="password"
              >
                PWD_LBL
              </label>
              <input
                autoComplete="new-password"
                className="input "
                id="password"
                name="password1"
                onChange={[Function]}
                pattern=".{6,}"
                placeholder="PWD_PHOLD"
                required={true}
                title="PWD_PTRN_DESC"
                type="password"
                value=""
              />
              <PasswordToggleIcon
                hidePassword={true}
                inputRef={
                  Object {
                    "current": null,
                  }
                }
                isVisible={false}
                parentClassName=""
                secondInputRef={
                  Object {
                    "current": null,
                  }
                }
                toggler={[Function]}
              />
            </div>
            <div
              className="row password-confirm"
            >
              <label
                htmlFor="password-confirm"
              >
                CONFIRM_PWD_LBL
              </label>
              <input
                autoComplete="new-password"
                className="input "
                id="password-confirm"
                name="password2"
                onChange={[Function]}
                pattern=".{6,}"
                placeholder="CONFIRM_PWD_PHOLD"
                required={true}
                title="PWD_PTRN_DESC"
                type="password"
                value=""
              />
              <PasswordToggleIcon
                hidePassword={true}
                inputRef={
                  Object {
                    "current": null,
                  }
                }
                isVisible={false}
                parentClassName=""
                secondInputRef={
                  Object {
                    "current": null,
                  }
                }
                toggler={[Function]}
              />
            </div>
          </div>
          <div
            className="row add-info"
          >
            REGISTER_ADD_INFO_TXT
          </div>
          <div
            className="row register"
          >
            <input
              className="button full"
              type="submit"
              value="REGISTER_BTN_TXT"
            />
          </div>
          <div
            className="row links"
          >
            <p>
              <Link
                className="link"
                to="/default/login"
              >
                LINKS_LOGIN_TXT
              </Link>
            </p>
          </div>
        </div>
      </form>
      <Connect(Contact) />
    </div>
  </div>
  <Route
    path="default/registration/:name"
    render={[Function]}
  />
</Fragment>
`;

exports[`Registration and Mobile Phone Verification interactions should render PhoneInput lazily and handlers should work correctly 1`] = `
<Suspense
  fallback={
    <input
      className="input"
      name="phone_number"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="enter mobile phone number"
      type="tel"
      value=""
    />
  }
>
  <lazy
    country="it"
    enableSearch={false}
    excludeCountries={Array []}
    inputProps={
      Object {
        "autoComplete": "tel",
        "className": "form-control input ",
        "id": "phone-number",
        "name": "phone_number",
        "required": true,
      }
    }
    name="phone_number"
    onChange={[Function]}
    onKeyDown={[Function]}
    onlyCountries={Array []}
    placeholder="enter mobile phone number"
    preferredCountries={Array []}
    value=""
  />
</Suspense>
`;
